### PR TITLE
Moved API to .rx subset and added in a classic way API on the top level

### PR DESCRIPTION
**Pull request by buh**
### Changes for Rx

Basically every rx API moved with `.rx` prefix.
```swift
Client.shared.rx.channel(query: channelQuery)
```

### A classic way API

It works with `Subscription` and `completion: ClientCompletion<T>` blocks.

A subscription for an async request.
You have to keep as an instance variable until you need results in a completion block.

For example:
```swift
class MyViewController: UIViewController {

    var subscription: Subscription?

    open override func viewDidLoad() {
        super.viewDidLoad()

        let channel = Channel(type: .messaging, id: "general")
        channel.create { result in
            let channelResponse = try result.get()
            print(channelResponse)
        }
    }
    open override func viewWillAppear() {
        super.viewWillAppear(animated)

        // Subscribe for Client events.
        subscription = Client.shared.onEvent() { event
            print(event)
        }
    }

    open override func viewWillDisappear() {
        super.viewWillDisappear()

        // Cancel subscription.
        subscription = nil
    }
}
```

### CODE REVIEW

File: Sources/Core/Client/API/Client+UnreadCount.swift [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+public typealias UnreadCount = (channels: Int, messages: Int)
+
+extension Client {
+    
+    /// Observe an unread count of messages in the channel.
+    /// - Note: Be sure the current user is a member of the channel.
+    /// - Note: 100 is the maximum unread count of messages.
+    /// - Parameter completion: a completion block with `UnreadCount`.
+    /// - Returns: a subscription.
+    public func unreadCount(_ completion: @escaping ClientCompletion<UnreadCount>) -> Subscription {
```
Comment (b-onc) [inline Sources/Core/Client/API/Client+UnreadCount.swift]: Is it possible `unreadCount` is a variable instead of a function with completion?
It doesn't make sense to
```
Client.shared.unreadCount { count in
  self.count = count
}
```
It makes sense, to me:
```
count = Client.shared.unreadCount
```

We can bind the `rx.unreadCount.asObservable()` to the variable?

File: Sources/Core/Client/API/Client+UnreadCount.swift [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        return rx.unreadCount.asObservable().bind(to: completion)
```
Comment (b-onc) [inline Sources/Core/Client/API/Client+UnreadCount.swift]: `bindOnce` instead of `bind`?

File: Sources/Core/Client/Client.swift [hunk 3]
```diff
@@ -156,7 +116,14 @@ public final class Client {
             Thread.callStackSymbols.forEach { ClientLogger.logger("", "", $0) }
         }
     }
-
+    
+    /// A subscription for websocket connection status.
+    /// - Parameter completion: a completion block (see `ClientCompletion`).
+    /// - Returns: a subscription.
+    public func connection(_ completion: @escaping ClientCompletion<WebSocket.Connection>) -> Subscription {
+        return rxConnection.bind(to: completion)
```
Comment (b-onc) [inline Sources/Core/Client/Client.swift]: `bindOnce` instead of `bind`?

File: Sources/Core/Client/WebSocket/WebSocket.swift [hunk 4]
```diff
@@ -73,6 +62,18 @@ public final class WebSocket {
         disconnect()
     }
     
+    /// Subscribe for websocket response.
+    /// - Parameter completion: a completion block (see `ClientCompletion`).
+    /// - Returns: a subscription.
+    public func response(_ completion: @escaping ClientCompletion<WebSocket.Response>) -> Subscription {
+        return rxResponse.bind(to: completion)
```
Comment (b-onc) [inline Sources/Core/Client/WebSocket/WebSocket.swift]: `bindOnce` instead of `bind`?

File: Sources/Core/Client/WebSocket/WebSocket.swift [hunk 5]
```diff
@@ -161,11 +177,87 @@ public final class WebSocket {
...
     }
     
-    func clearStateAfterDisconnect() {
+    private func isTokenExpired() -> Bool {
+        guard let lastJSONError = lastJSONError else {
+            return false
+        }
+        
+        let isTokenExpired = lastJSONError.code == ClientErrorResponse.tokenExpiredErrorCode
+        return isTokenExpired && Client.shared.touchTokenProvider()
```
Comment (b-onc) [inline Sources/Core/Client/WebSocket/WebSocket.swift]: We really need dependency injection.. `WebSocket` instance method mutating singleton `Client` is unexpected

File: Sources/Core/Client/WebSocket/WebSocket.swift [hunk 6]
```diff
@@ -161,11 +177,87 @@ public final class WebSocket {
...
+        return .notConnected
+    }
+    
+    private func willReconnectAfterError(_ error: Swift.Error) -> Bool {
+        if isStopError(error) {
+            return false
+        }
+        
+        if InternetConnection.shared.isAvailable {
+            reconnect()
```
Comment (b-onc) [inline Sources/Core/Client/WebSocket/WebSocket.swift]: `willReconnectAfterError` calling `reconnect` is unexpected, to me it doesn't sound like a mutating func

File: Sources/Core/Model/Channel/Channel+Events.swift [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+public extension Channel {
+    
+    /// Observe channel events.
+    /// - Parameters:
+    ///   - eventType: an event type.
+    ///   - completion: a completion block with `Event`.
+    /// - Returns: a subscription.
+    func onEvent(_ eventType: EventType, _ completion: @escaping ClientCompletion<Event>) -> Subscription {
+        return rx.onEvent(eventType).bind(to: completion)
```
Comment (b-onc) [inline Sources/Core/Model/Channel/Channel+Events.swift]: `bindOnce` instead of `bind` (to all functions in this file)?

File: Sources/Core/Model/Message/Message.swift [hunk 8]
```diff
@@ -30,6 +30,8 @@ public struct Message: Codable {
         case reactionCounts = "reaction_counts"
     }
     
+    static var flaggedIds = [String]()
```
Comment (b-onc) [inline Sources/Core/Model/Message/Message.swift]: This can be a `Set` to make `contains(id)` operation O(1) instead of O(n)

File: Sources/Core/Client/API/Client+UnreadCount.swift (hunk 1 above)
Comment (buh) [reply Sources/Core/Client/API/Client+UnreadCount.swift]: It returns `Subscription` for getting continuously values. it's not a variable.
If you need the current value you have `Channel.currentUnreadCount` variable.

File: Sources/Core/Client/API/Client+UnreadCount.swift (hunk 2 above)
Comment (buh) [reply Sources/Core/Client/API/Client+UnreadCount.swift]: It returns `Subscription` because it continuously will values.

File: Sources/Core/Client/Client.swift (hunk 3 above)
Comment (buh) [reply Sources/Core/Client/Client.swift]: Nope. You are observing the connection state continuously.

File: Sources/Core/Client/WebSocket/WebSocket.swift (hunk 4 above)
Comment (buh) [reply Sources/Core/Client/WebSocket/WebSocket.swift]: Nope. You are observing events continuously.

File: Sources/Core/Client/WebSocket/WebSocket.swift (hunk 5 above)
Comment (buh) [reply Sources/Core/Client/WebSocket/WebSocket.swift]: How do you suggest to inject it?

File: Sources/Core/Client/WebSocket/WebSocket.swift (hunk 6 above)
Comment (buh) [reply Sources/Core/Client/WebSocket/WebSocket.swift]: I renamed it.

File: Sources/Core/Model/Channel/Channel+Events.swift (hunk 7 above)
Comment (buh) [reply Sources/Core/Model/Channel/Channel+Events.swift]: Nope. You are observing events continuously.

File: Sources/Core/Client/API/Client+UnreadCount.swift (hunk 1 above)
Comment (b-onc) [reply Sources/Core/Client/API/Client+UnreadCount.swift]: Then we should rename `completion` to `onChanged` or similar IMHO

File: Sources/Core/Client/WebSocket/WebSocket.swift (hunk 5 above)
Comment (b-onc) [reply Sources/Core/Client/WebSocket/WebSocket.swift]: Initializer?

File: Sources/Core/Model/Message/Message+RxRequests.swift [hunk 9]
```diff
@@ -85,51 +80,7 @@ public extension Message {
     }
     
     private func flagUnflagMessage(endpoint: Endpoint) -> Observable<FlagMessageResponse> {
-        return Client.shared.flagUnflag(endpoint: endpoint,
-                                        aleradyFlagged: FlagMessageResponse(messageId: id, created: Date(), updated: Date()))
+        let aleradyFlagged = FlagMessageResponse(messageId: base.id, created: Date(), updated: Date())
```
Comment (b-onc) [inline Sources/Core/Model/Message/Message+RxRequests.swift]: Typo in `aleradyFlagged`

File: Sources/Core/Client/WebSocket/WebSocket.swift (hunk 5 above)
Comment (buh) [reply Sources/Core/Client/WebSocket/WebSocket.swift]: `Client` is a shared instance.

File: Sources/Core/Model/User+Requests.swift [hunk 10]
```diff
@@ -91,8 +91,8 @@ public extension User {
     }
     
     private func flagUnflagUser(endpoint: Endpoint) -> Observable<FlagUserResponse> {
-        return Client.shared.flagUnflag(endpoint: endpoint,
-                                        aleradyFlagged: FlagUserResponse(user: self, created: Date(), updated: Date()))
+        return Client.shared.rx.flagUnflag(endpoint: endpoint,
+                                           aleradyFlagged: FlagUserResponse(user: self, created: Date(), updated: Date()))
```
Comment (b-onc) [inline Sources/Core/Model/User+Requests.swift]: Typo in `aleradyFlagged`

File: Sources/Core/Client/WebSocket/WebSocket.swift (hunk 5 above)
Comment (buh) [reply Sources/Core/Client/WebSocket/WebSocket.swift]: Why do you want to inject WebSocket?

Comment (b-onc) [reply Sources/Core/Client/WebSocket/WebSocket.swift]: I want to inject `Client` into `WebSocket` (or anywhere where it's used) so that testing is easier and it's easier to follow & understand. Currently, `Client.shared` is accessed from multiple places which makes it hard to follow where it's mutated from.

Comment (b-onc) [reply Sources/Core/Client/WebSocket/WebSocket.swift]: It's probably not for this PR, though, out of scope

Comment (buh) [reply Sources/Core/Client/WebSocket/WebSocket.swift]: `Client` use `WebSocket`. `WebSocket` _doesn't know_ about the `Client`.

Comment (b-onc) [reply Sources/Core/Client/WebSocket/WebSocket.swift]: It does. Here, in an instance method of `WebSocket`, `Client.shared` is accessed

Comment (buh) [reply Sources/Core/Client/WebSocket/WebSocket.swift]: It's only used for logger, that easy to avoid. I mean in general functionality.
